### PR TITLE
[Tutorial] Single Machine Multi-GPU Minibatch Graph Classification

### PR DESCRIPTION
**Pull request by mufeili**
## Description
Currently building the doc will not trigger the multi-gpu training due to `if __name__ == '__main__'`. I'm not sure if we want to enable that for multi-gpu tutorials.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tutorials/multi/1_graph_classification.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
+"""
+Single Machine Multi-GPU Minibatch Graph Classification
+=======================================================
+
+In this tutorial, you will learn how to use multiple GPUs in training a
+graph neural network (GNN) for graph classification. This tutorial assumes
+expertise in GNNs for graph classification and we recommend you to check
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:7]:
```suggestion
familiarity in GNNs for graph classification and we recommend you to check
```

File: tutorials/multi/1_graph_classification.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+"""
+
+###############################################################################
+# The node and edge features in the graphs, if any, will also be on the GPU.
+# After that, the forward computation, backward computation and parameter
+# update will take place on the GPU. For graph classification, this repeats
+# for each minibatch gradient descent.
+#
+# Using multiple GPUs allows performing more computation per unit of time. It
+# is like having a team work together, where each GPU is a team member. We need
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:33]: `It is like` grammar error
-> It works like a team

File: tutorials/multi/1_graph_classification.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        dist.barrier()
+
+        val_acc = evaluate(model, val_loader, device)
+        print('Val acc: {:.4f}'.format(val_acc))
+        dist.barrier()
+
+    test_acc = evaluate(model, test_loader, device)
+    print('Test acc: {:.4f}'.format(test_acc))
+    dist.barrier()
+    dist.destroy_process_group()
```
Comment (hetong007) [inline tutorials/multi/1_graph_classification.py:222]: Why do we destroy in `main` instead of `'__main__'`?

File: tutorials/multi/1_graph_classification.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+###############################################################################
+# The node and edge features in the graphs, if any, will also be on the GPU.
+# After that, the forward computation, backward computation and parameter
+# update will take place on the GPU. For graph classification, this repeats
+# for each minibatch gradient descent.
+#
+# Using multiple GPUs allows performing more computation per unit of time. It
+# is like having a team work together, where each GPU is a team member. We need
+# to distribute the computation workload across GPUs and let them synchronize
+# the efforts regularly. PyTorch provides convenient APIs for this task with
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:35]: `with multiple processes` -> `by multiple processes`?
Also you can add a link to pytorch's multigpu tutorial here

**Review by hetong007 — APPROVED** [review]
In general LGTM.

File: tutorials/multi/1_graph_classification.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+# multiple processes, one per GPU, and we can use them in conjunction with DGL.
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:36]: `and we can use them in conjunction with DGL.` -> We follow those standard practices in DGL also.?

File: tutorials/multi/1_graph_classification.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+# allows multiple GPUs to perform the forward and backward computation of
+# multiple gradient descents in parallel. To distribute a dataset across
+# multiple GPUs, we need to partition it into multiple mutually exclusive
+# subsets of a similar size, one per GPU. We need to repeat the random
+# partition every epoch to guarantee randomness. We can use
+# :func:`~dgl.dataloading.pytorch.GraphDataLoader`, which wraps some PyTorch 
+# APIs and does the job for graph classification in data loading.
+#
+# Once all GPUs have finished the backward computation for its minibatch,
+# we need to synchronize the model parameter update across them. Specifically,
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:48]: > we need to synchronize the model parameter update across them

vs

> the model parameters need to be synchronized update across them

which one is better?

File: tutorials/multi/1_graph_classification.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+# this involves collecting gradients from all GPUs, averaging them and updating
+# the model parameters on each GPU. We can wrap a PyTorch model with
+# :func:`~torch.nn.parallel.DistributedDataParallel` so that the model
+# parameter update will invoke gradient synchronization first under the hood.
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:52]: What does `first` mean here?

File: tutorials/multi/1_graph_classification.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+#
+# That’s the core behind this tutorial. We will explore it in more detail with
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:54]: explore it more in detail

File: tutorials/multi/1_graph_classification.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        backend='nccl',
+        init_method='tcp://127.0.0.1:12345',
+        world_size=world_size,
+        rank=rank)
+
+###############################################################################
+# Data Loader Preparation
+# -----------------------
+#
+# Split the dataset into training, validation and test subsets. In dataset
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:79]: The first sentence seems incomplete

File: tutorials/multi/1_graph_classification.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+    def forward(self, g, feats):
+        feats = self.conv1(g, feats)
+        feats = F.relu(feats)
+        feats = self.conv2(g, feats)
+
+        return self.pool(g, feats)
+
+###############################################################################
+# To ensure same initial model parameters across processes, we need to set a
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:129]: `To ensure same initial model parameters across processes` read weird

Comment (VoVAllen) [reply tutorials/multi/1_graph_classification.py:129]: set a same random seed -> set the same random seed

File: tutorials/multi/1_graph_classification.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+# same random seed before model initialization. Once constructed, we wrap the
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:130]: if use `constructed` here, the subjective of the sentence should be model?

File: tutorials/multi/1_graph_classification.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+def init_model(seed, device):
+    torch.manual_seed(seed)
+    model = GIN().to(device)
+    model = DistributedDataParallel(model, device_ids=[device], output_device=device)
+
+    return model
+
+###############################################################################
+# Main Function for One Process
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:145]: for each/all process?

File: tutorials/multi/1_graph_classification.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        with torch.no_grad():
+            pred = model(bg, feats)
+        _, pred = torch.max(pred, 1)
+        total += len(labels)
+        total_correct += (pred == labels).sum().cpu().item()
+
+    return 1.0 * total_correct / total
+
+###############################################################################
+# Define the main function for one process. Since different processes handle
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:171]: for each/all process?

File: tutorials/multi/1_graph_classification.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+# different data batches, some of them can finish their task earlier.
+# `dist.barrier()` forces faster processes to wait until all processes
+# have encountered the barrier.
```
Comment (VoVAllen) [inline tutorials/multi/1_graph_classification.py:174]: `encountered the barrier` -> `reached the barrier`

File: tutorials/multi/1_graph_classification.py (hunk 11 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:130]: Yes.

File: tutorials/multi/1_graph_classification.py (hunk 10 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:129]: It's just ensure sth.

File: tutorials/multi/1_graph_classification.py (hunk 9 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:79]: It's an imperative sentence. See https://literarydevices.net/imperative-sentence/#:~:text=An%20imperative%20sentence%20is%20a,exclamation%20mark%20or%20a%20period.

File: tutorials/multi/1_graph_classification.py (hunk 7 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:52]: before actually updating the model parameters

File: tutorials/multi/1_graph_classification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:7]: Changed to knowledge.

File: tutorials/multi/1_graph_classification.py (hunk 6 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:48]: I'll stick to the first one.

File: tutorials/multi/1_graph_classification.py (hunk 5 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:36]: If we want to say "practices", then these are broader than particular APIs.

File: tutorials/multi/1_graph_classification.py (hunk 4 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:35]: I cannot find results with "by multiple processes" in the search engine, but I can find ones with "with multiple processes" or "using multiple processes".

File: tutorials/multi/1_graph_classification.py (hunk 2 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:33]: "Be like" is a valid usage. See https://en.wiktionary.org/wiki/be_like.

File: tutorials/multi/1_graph_classification.py (hunk 11 above)
Comment (VoVAllen) [reply tutorials/multi/1_graph_classification.py:130]: I mean it should be `once constructed, model blablabla` instead of `once constructed, we blablabla`?

File: tutorials/multi/1_graph_classification.py (hunk 5 above)
Comment (VoVAllen) [reply tutorials/multi/1_graph_classification.py:36]: I just feel you can start another sentence for the `we can use them in conjunction with DGL`. And `in conjunction with DGL` reads tedious to me

File: tutorials/multi/1_graph_classification.py (hunk 9 above)
Comment (VoVAllen) [reply tutorials/multi/1_graph_classification.py:79]: I know it's valid in grammar. But I'm wondering why you change the way of using `we` as the subjective here. I personally feel imperative sentence here breaks the whole flow in the tutorial, the tone here is harder than the other part

File: tutorials/multi/1_graph_classification.py (hunk 10 above)
Comment (VoVAllen) [reply tutorials/multi/1_graph_classification.py]: But you are ensure the sameness instead of the model parameters here?

File: tutorials/multi/1_graph_classification.py [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+# Intuitively, we can distribute the workload along the dimension of data. This
+# allows multiple GPUs to perform the forward and backward computation of
+# multiple gradient descents in parallel. To distribute a dataset across
+# multiple GPUs, we need to partition it into multiple mutually exclusive
+# subsets of a similar size, one per GPU. We need to repeat the random
+# partition every epoch to guarantee randomness. We can use
+# :func:`~dgl.dataloading.pytorch.GraphDataLoader`, which wraps some PyTorch 
+# APIs and does the job for graph classification in data loading.
+#
+# Once all GPUs have finished the backward computation for its minibatch,
```
[the remark this answers is not in the document]
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:47]: See if the figure here (http://s3.us-west-2.amazonaws.com/dgl-data/tutorial/mgpu_gc.png) looks good to you.

File: tutorials/multi/1_graph_classification.py [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+            pred = model(bg, feats)
+
+            loss = criterion(pred, labels)
+            total_loss += loss.cpu().item()
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+        loss = total_loss
+        print('Loss: {:.4f}'.format(loss))
+        dist.barrier()
```
[the remark this answers is not in the document]
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:213]: In most cases, it should work fine without the barrier.

File: tutorials/multi/1_graph_classification.py (hunk 3 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:222]: Honestly I'm not sure and simply following the practice in the PyTorch tutorial [here](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html). This might be related to that `init_process_group` is called per process.

File: tutorials/multi/1_graph_classification.py (hunk 5 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:36]: How about ", and we can use them in conjunction with DGL." -> ". They are compatible with DGL."?

File: tutorials/multi/1_graph_classification.py (hunk 9 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:79]: How about adding "We" at the beginning?

File: tutorials/multi/1_graph_classification.py (hunk 10 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py]: We are ensuring same model parameters. Sth with an adjective is still sth.

File: tutorials/multi/1_graph_classification.py (hunk 11 above)
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:130]: Changed to "Once we construct a model instance,"

File: tutorials/multi/1_graph_classification.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (hetong007) [reply tutorials/multi/1_graph_classification.py:47]: Looks good in general. One question though: why are the blocks "local gradient" having different bg colors?

File: tutorials/multi/1_graph_classification.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (hetong007) [reply tutorials/multi/1_graph_classification.py:217]: Then, what are the cases that one needs to add it? My concern would be that the readers are all gonna follow it, and might not be the best practice as it is not necessary.

File: tutorials/multi/1_graph_classification.py (hunk 3 above)
Comment (hetong007) [reply tutorials/multi/1_graph_classification.py:226]: OK that's fair.

File: tutorials/multi/1_graph_classification.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:47]: To indicate that they are different

File: tutorials/multi/1_graph_classification.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply tutorials/multi/1_graph_classification.py:217]: How about simply removing them? As you said, they are unnecessary in this example. To fully explain when this will be necessary might be out of the scope of this tutorial.

[the remark this answers is not in the document]
Comment (hetong007) [reply tutorials/multi/1_graph_classification.py:217]: OK let's remove it.

**Review by VoVAllen — APPROVED** [review]
LGTM. However, I still feel uncomfortable with the `ensure` usage. Would like to hear from other people on this point

File: tutorials/multi/1_graph_classification.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (hetong007) [reply tutorials/multi/1_graph_classification.py:47]: In this case, we might have the mini-batch gradient having the average color of the two.

[the remark this answers is not in the document]
Comment (hetong007) [reply tutorials/multi/1_graph_classification.py:47]: And the similar blue-ish shadow is shared between model and gradient, maybe also make them more different from each other?
